### PR TITLE
Add functions to lattice and accelerator

### PR DESCRIPTION
**Pull request by fernandohds564**
This PR adds the following features:

 - Create method `insert` in `Accelerator` object. Allows
```python3
mod.insert(554, ele)
```
 - Create method `__radd__` for `Accelerator` object. Allows: 
```python3
mod2 = ele + mod
```
 - Create methods `__mul__` for `Accelerator` object. Allows: 
```python3
mod2 = mod * 5
```
 - Generalize function shift to also handle negative start values: 
```python3
(pyaccel.lattice.shift(mod, -345) == pyaccel.lattice.shift(mod, len(mod)-345)) is True
```
 - Create function that split `Element` object into several fractions:
```python3
elems = pyaccel.lattice.split_element(ele, [0.25, 0.35, 0.4])
```
 - Create function to insert element at a given longitudinal position of the accelerator model: 
```python3
mod2 = pyaccel.lattice.insert_marker_at_position(mod, 235.64)
```

The last function proved to be very useful when we want to get optics properties, such as beam stay clear (BSC) or beta functions at a given position of the lattice.
Follow below an example script which makes use of this function.
I created this script to study the effect of the Delta22 vacuum chamber on the BSC of the storage ring.

Running:
```python3
from pymodels import si
import pyaccel
import numpy as np 
import matplotlib.pyplot as mplt
from matplotlib.patches import Ellipse, Rectangle


def calc_bsc(dlt22_chamb=False, chamb_side=6.4, chamb_len=2.6, plot=False):

    mod = si.create_accelerator(ids_vchamber=True)

    mib = pyaccel.lattice.find_indices(mod, 'fam_name', 'mib')
    spos = pyaccel.lattice.find_spos(mod)
    mod = pyaccel.lattice.insert_marker_at_position(
        mod, 'end_delta22', spos[mib[0]] + chamb_len/2)
    mod = pyaccel.lattice.insert_marker_at_position(
        mod, 'end_delta22', spos[mib[0]] - chamb_len/2)
    
    dlt22 = pyaccel.lattice.find_indices(
        mod, 'fam_name', 'end_delta22')

    accepx, accepy, twiss = pyaccel.optics.calc_transverse_acceptance(mod)
    accepx = np.min(accepx)
    accepy = np.min(accepy)
    bscx = np.sqrt(twiss.betax*accepx)
    bscy = np.sqrt(twiss.betay*accepy)

    # point where diamond shaped chamber crosses xy axes:
    d = chamb_side/np.sqrt(2)  
    
    if dlt22_chamb:
        # Reduce horizontal aperture to not compromise vertical BSC:
        a = bscy[dlt22[0]]
        b = np.sqrt(d*d*1e-6 - a*a)

        for i in range(dlt22[0], dlt22[1]+1):
            mod[i].hmin = -b
            mod[i].hmax = b
            mod[i].vmin = -a
            mod[i].vmax = a
           
        accepx, accepy, twiss = pyaccel.optics.calc_transverse_acceptance(mod)
        accepx = np.min(accepx)
        accepy = np.min(accepy)
        bscx = np.sqrt(twiss.betax*accepx)
        bscy = np.sqrt(twiss.betay*accepy)

    if plot:        
        fig, ax = mplt.subplots(1, 1, figsize=(5, 5))
        bscxi = bscx[dlt22[1]] * 1e3
        bscyi = bscy[dlt22[1]] * 1e3
        ell = Ellipse(
            xy=(0, 0), width=2*bscxi, height=2*bscyi, fill=False,
            ec='tab:blue', lw=2, label='BSC @ end of Delta22')
        ax.add_patch(ell)

        square_rotx = d * np.array([1, 0, -1, 0, 1])
        square_roty = d * np.array([0, 1, 0, -1, 0])

        modx = 1e3*b * np.array([1, -1, -1, 1, 1])
        mody = 1e3*a * np.array([1, 1, -1, -1, 1])

        tit = 'Without Delta22'
        if dlt22_chamb:
            tit = 'With Delta22'
        
        ax.set_title('Beam Stay Clear for Machine ' + tit)
        ax.plot(square_rotx, square_roty, lw=2, color='k', label='Chamber')
        ax.plot(modx, mody, 'k--', lw=2, label='Model of the Chamber')
        ax.set_xlim([-d*1.1, d*1.1])
        ax.set_ylim([-d*1.1, d*1.1])
        ax.set_xlabel('X [mm]')
        ax.set_ylabel('Y [mm]')
        ax.legend(loc='best')
        ax.grid(True, alpha=0.5, ls='--', lw=1)
        fig.tight_layout()
        ax.set_aspect('equal')

    hmax = pyaccel.lattice.get_attribute(mod, 'hmax')
    vmax = pyaccel.lattice.get_attribute(mod, 'vmax')
    return mod, twiss, bscx, bscy, hmax, vmax

plot_beta = False
modn, twissn, bscxn, bscyn, hmaxn, vmaxn = calc_bsc(dlt22_chamb=False)
mod, twiss, bscx, bscy, hmax, vmax = calc_bsc(dlt22_chamb=True, plot=True)
pos = twiss.spos

if plot_beta:
    fig, (ax, ay) = mplt.subplots(2, 1, figsize=(10, 9), sharex=True)
    ay.plot(pos, twiss.betax, color='tab:blue', label='Horizontal')
    ay.plot(pos, twiss.betay, color='tab:red', label='Vertical')
    ay.grid(True, alpha=0.5, ls='--', lw=1)
    ay.set_ylabel('Beta [m]', fontsize=14)
    ay.set_xlabel('Longitudinal Position [m]', fontsize=14)
    ay.legend(loc='best')
else:
    fig, ax = mplt.subplots(1, 1, figsize=(10, 5))
    ax.set_xlabel('Longitudinal Position [m]', fontsize=14)

ax.plot(pos, bscxn*1e3, color='tab:blue', label='Hor. BSC W/o Delta22')
ax.plot(pos, -bscyn*1e3, color='tab:red', label='-Vert. BSC W/o Delta22')
ax.plot(pos, bscx*1e3, color='tab:cyan', label='Hor. BSC With Delta22')
ax.plot(pos, -bscy*1e3, color='orange', label='-Vert. BSC With Delta22')
pos = pos[:-1]
ax.plot(pos, hmax*1e3, 'k', label='Hor. Chamb. With Delta22 @ 02SB')
ax.plot(pos, -vmax*1e3, 'k--', label='-Vert. Chamb. With Delta22 @ 02SB')

ax.grid(True, alpha=0.5, ls='--', lw=1)
ax.set_ylabel('Beam Stay Clear [mm]', fontsize=14)
ax.legend(loc='lower center', bbox_to_anchor=(0.5, 1), ncol=3)
pyaccel.graphics.draw_lattice(mod, gca=ax, height=1.5)
ax.set_xlim([0, mod.length/5])

fig.tight_layout()
mplt.show()
```

Will result in:
image

and:
image

### CODE REVIEW

File: pyaccel/accelerator.py [hunk 1]
```diff
@@ -162,11 +162,42 @@ def pop(self, index):
...
+    def insert(self, index: int, element: _elements.Element):
+        """Insert element at a given index of the lattice.
+
+        Args:
+            index (int): index where to insert element. May be negative or
+                positive
+            element (pyaccel.elements.Element): Element to be inserted.
+
+        Raises:
+            TypeError: when element is no an pyaccel.elements.Element.
```
Comment (anacso17) [inline pyaccel/accelerator.py:188]: "..not a pyaccel.."?

File: pyaccel/lattice.py [hunk 2]
```diff
@@ -40,27 +40,183 @@ def build(elist):
...
 
 
+@_interactive
+def insert_marker_at_position(accelerator, fam_name: str, position: float):
+    """Return a copy of accelerator with a marker at the desired position.
+
+    This method will split a element of the accelerator if needed to ensure
+    the position of the marker is respected.
+    The vacuum chamber of the marker will be defined by the element closest to
+    in in the model.
```
Comment (anacso17) [inline pyaccel/lattice.py:72]: "..closest to it.."?

**Review by anacso17 — APPROVED** [review]

**Review by vellosok75 — APPROVED** [review]

File: pyaccel/accelerator.py (hunk 1 above)
Comment (fernandohds564) [reply pyaccel/accelerator.py:188]: Thx @anacso17, done!

File: pyaccel/lattice.py (hunk 2 above)
Comment (fernandohds564) [reply pyaccel/lattice.py:72]: Thx @anacso17, done!

**Review by murilobalves — APPROVED** [review]

**Review by LuanaVilela — APPROVED** [review]

**Review by xresende — APPROVED** [review]

**Pull request merged**